### PR TITLE
627 New 'geoips tree' command

### PR DESCRIPTION
**Pull request by evrose54**
# Reviewer Checklist

* [x] Required ***existing tests*** pass (ie full_test.sh, others as appropriate)
* [x] Required ***unit tests*** added and pass for new/modified functionality
* [x] NO REQUIRED ***integration tests*** (explain why not required)
* [x] Required ***documentation*** added for new/modified functionality
* [x] Required ***release notes*** added for new/modified functionality
* [x] NO REQUIRED ***updates to other repos*** (explain why not required)

https://github.com/NRLMMD-GEOIPS/.github/blob/main/.github/review-template.md

# Related Issues
fixes NRLMMD-GEOIPS/geoips#627

# Testing Instructions
Run pytest -v ./tests/unit_tests/commandline/test_geoips_tree.py.

# Summary
The GeoIPS CLI provides a variety of commands which aren't necessarily easily exposed
via ``geoips -h``. To improve this issue, we've added a ``geoips tree`` command which
exposes all GeoIPS CLI commands in a tree-like fashion. This way, we can expose all
commands that are available via the GeoIPS CLI, and expose the depth in which these
commands exist.

By displaying the commands in a depthwise structure, users can understand what commands
are available and how they are called.

If you just call ``geoips tree``, you'll get the full command tree in a non-colored,
verbose output.

The output of running ``geoips tree`` is shown below.

```
    geoips tree

    geoips
        geoips config
            geoips config install
        geoips get
            geoips get family
            geoips get interface
            geoips get package
            geoips get plugin
        geoips list
            geoips list interface
            geoips list interfaces
            geoips list packages
            geoips list plugins
            geoips list scripts
            geoips list test-datasets
            geoips list unit-tests
        geoips run
            geoips run single_source
            geoips run data_fusion
            geoips run config_based
        geoips test
            geoips test linting
            geoips test script
        geoips tree
        geoips validate
```

``geoips tree`` additionaly provides optional arguments to filter the output of this
command. Shown below are these optional arguments and descriptions of what each argument
does.

* ``--colored``

  * The output of ``geoips tree`` might be a little hard to interpret. If you want the
    output of ``geoips tree`` to be colored by depth, make sure to use the ``--colored``
    flag. (Defaults to False)

* ``--max_depth``

  * How many levels of the tree we'd like to expose. Defaults to two levels, which is
    shown above.

* ``--short_name``

  * The output of ``geoips tree`` provides the full command string at each level. If you
    just want the literal command name and every level, make sure to provide this flag.
    (Defaults to False)

```
    modified: docs/source/userguide/command_line.rst
    modified: geoips/commandline/ancillary_info/cmd_instructions.yaml
    modified: geoips/commandline/commandline_interface.py
    modified: geoips/commandline/geoips_command.py
    modified: tests/unit_tests/commandline/cli_top_level_tester.py
    added: geoips/commandline/geoips_tree.py
    added: tests/unit_tests/commandline/test_geoips_tree.py
```

# Output
Screenshot 2024-06-10 at 11 14 18 AM

### CODE REVIEW

**Comment by evrose54** [comment]
Note: Once the updated documentation is actually up on GitHub.io the hyperlinked outputs of the tree will point towards the appropriate location in the documentation.

**Comment by biosafetylvl5** [comment]
Could we format like this? 
```
    geoips tree
        geoips config [install]
        geoips get [family, interface, package, plugin]
        geoips list [interface, interfaces, packages, plugins, scripts, test-datasets, unit-tests]
        geoips run [single_source, data_fusion, config_based]
        geoips test [linting, script]
        geoips tree
        geoips validate
     ```

File: geoips/commandline/geoips_tree.py
Comment (jsolbrig) [inline geoips/commandline/geoips_tree.py:82]: I don't think that we need choices here. Just accept any integer. If the integer is greater than the number of available levels, just print what is available. So, `--max-depth 4` should result in the same output as `--max-depth 2` given the current CLI.

File: geoips/commandline/geoips_tree.py
Comment (jsolbrig) [inline geoips/commandline/geoips_tree.py:92]: Dash, not underscore.

**Review by jsolbrig — CHANGES_REQUESTED** [review]
Please address the comments.

File: geoips/commandline/geoips_tree.py
Comment (evrose54) [reply geoips/commandline/geoips_tree.py]: It's a dash now!

File: geoips/commandline/geoips_tree.py
Comment (evrose54) [reply geoips/commandline/geoips_tree.py]: It's ``--color`` now.

File: geoips/commandline/geoips_tree.py
Comment (evrose54) [reply geoips/commandline/geoips_tree.py]: It's ``--max-depth`` now.

File: geoips/commandline/geoips_tree.py
Comment (evrose54) [reply geoips/commandline/geoips_tree.py]: Removed choices and added a check in the call function to confirm that the max depth is in the valid range.

**Review by mindyls — DISMISSED** [review]
I left a few comments, leave replies and ping me to resolve when done!

File: docs/source/userguide/command_line.rst
Comment (evrose54) [reply docs/source/userguide/command_line.rst:622]: I think we agreed for no color to be the default on this one. I know we spoke about this when reviewing Gwyn's automated release note code, but I think it makes more sense for this command to not output color as a default.

File: docs/source/userguide/command_line.rst
Comment (evrose54) [reply docs/source/userguide/command_line.rst:404]: Yes, I think that will be done in a separate PR. Still need to decide on whether or not we want to support plugin-package installation with the CLI or if we should leave that to git and/or pip.

File: geoips/commandline/ancillary_info/cmd_instructions.yaml
Comment (evrose54) [reply geoips/commandline/ancillary_info/cmd_instructions.yaml:296]: I can expand on that point!

File: geoips/commandline/geoips_command.py
Comment (evrose54) [reply geoips/commandline/geoips_command.py:45]: Yes, this is done in v1.13 release PR but the CLI fails with the old implementation shown above. I had to add that as those changes weren't included in this branch. Shouldn't cause any conflicts though, as it's the exact same code as what is implemented in v1.13 release.

File: geoips/commandline/geoips_tree.py
Comment (evrose54) [reply geoips/commandline/geoips_tree.py:15]: Will do!

File: geoips/commandline/geoips_tree.py
Comment (evrose54) [reply geoips/commandline/geoips_tree.py:82]: Will do!

File: tests/unit_tests/commandline/cli_top_level_tester.py
Comment (evrose54) [reply tests/unit_tests/commandline/cli_top_level_tester.py:68]: See the comment I left addressing this type of functionality. I know the changes in ``geoips_command`` are made in v1.13 release PR however I'm not sure if this change was actually included in that PR. I think this update was necessary for the unit tests to run.

**Comment by evrose54** [comment]
We should ignore ``Old docs have not been changed`` CI check as we need this documentation and will likely move all CLI documentation over to new docs in a single PR.

File: geoips/commandline/geoips_tree.py
Comment (mindyls) [inline geoips/commandline/geoips_tree.py:47]: Should this be set somewhere else besides in this command line script?  An environment variable -> base_paths.py which will eventually be in a config ?

Comment (mindyls) [reply geoips/commandline/geoips_tree.py:47]: Oh yeah, for now we can probably set 

gpaths["GEOIPS_DOCS_URL"] or something?  In base_paths?

File: geoips/commandline/geoips_tree.py
Comment (jsolbrig) [inline geoips/commandline/geoips_tree.py:3]: This should be `--max-depth` here.

File: docs/source/releases/latest/627-CLI-tree.yaml
Comment (jsolbrig) [inline docs/source/releases/latest/627-CLI-tree.yaml:59]: This should be `--max-depth`

File: docs/source/userguide/command_line.rst
Comment (jsolbrig) [inline docs/source/userguide/command_line.rst:616]: This should be `--max-depth`

File: docs/source/releases/latest/627-CLI-tree.yaml
Comment (jsolbrig) [inline docs/source/releases/latest/627-CLI-tree.yaml:64]: This should be `--short-name`

File: docs/source/userguide/command_line.rst
Comment (jsolbrig) [inline docs/source/userguide/command_line.rst:621]: This should be `--short-name`

File: geoips/commandline/ancillary_info/cmd_instructions.yaml
Comment (jsolbrig) [inline geoips/commandline/ancillary_info/cmd_instructions.yaml:303]: Nitpick: "Ie." should have two periods in it like this: "I.e.". It stands for "id est" which is Latin for "that is".

Comment (jsolbrig) [inline geoips/commandline/ancillary_info/cmd_instructions.yaml:303]: Using `--max-depth 0` doesn't work.

```bash
> geoips tree --max-depth 0

usage: To use, type `geoips tree <--max-depth> <int> <--color> <--short-name>`.
geoips tree: error: Invalid depth value: 0. Must be greater than or equal to 1.
```

Comment (evrose54) [reply geoips/commandline/ancillary_info/cmd_instructions.yaml:303]: This is fixed now.

File: geoips/commandline/geoips_tree.py
Comment (evrose54) [reply geoips/commandline/geoips_tree.py:148]: ``r"https://nrlmmd-geoips.github.io/geoips/userguide/command_line.html"``

Comment (evrose54) [reply geoips/commandline/geoips_tree.py:148]: It's an attribute of ``GeoipsTree``

File: geoips/commandline/geoips_tree.py
Comment (evrose54) [reply geoips/commandline/geoips_tree.py:47]: I'll make that change.

File: geoips/commandline/geoips_tree.py
Comment (evrose54) [reply geoips/commandline/geoips_tree.py:3]: Fixed!

File: docs/source/releases/latest/627-CLI-tree.yaml
Comment (evrose54) [reply docs/source/releases/latest/627-CLI-tree.yaml:59]: Fixed!

File: docs/source/userguide/command_line.rst
Comment (evrose54) [reply docs/source/userguide/command_line.rst:616]: Fixed!

File: docs/source/userguide/command_line.rst
Comment (evrose54) [reply docs/source/userguide/command_line.rst:621]: Fixed!

File: geoips/commandline/ancillary_info/cmd_instructions.yaml
Comment (evrose54) [reply geoips/commandline/ancillary_info/cmd_instructions.yaml:303]: Fixed!

**Review by jsolbrig — APPROVED** [review]

**Review by jsolbrig — APPROVED** [review]

File: docs/source/releases/latest/627-CLI-tree.yaml
Comment (mindyls) [inline docs/source/releases/latest/627-CLI-tree.yaml:32]: Note this is different now.

File: docs/source/userguide/command_line.rst
Comment (mindyls) [inline docs/source/userguide/command_line.rst:93]: This has changed now

Comment (mindyls) [reply docs/source/userguide/command_line.rst:93]: Or maybe not yet?

File: docs/source/userguide/command_line.rst
Comment (mindyls) [inline docs/source/userguide/command_line.rst:185]: This has changed now, though, right?

File: docs/source/userguide/command_line.rst
Comment (mindyls) [inline docs/source/userguide/command_line.rst:590]: This has changed.

File: docs/source/userguide/command_line.rst
Comment (mindyls) [inline docs/source/userguide/command_line.rst:611]: This was changed to --color at some point right?

File: docs/source/releases/latest/627-CLI-tree.yaml
Comment (evrose54) [reply docs/source/releases/latest/627-CLI-tree.yaml:32]: Ah; yes it is. Is it worth it to change this in this PR or should we do one final swoop at the CLI documentation once the other CLI PRs are merged. Since this is going to have to be hardcoded, it might just be best to wait fixing this until all new commands are merged. Up to you though.

File: docs/source/userguide/command_line.rst
Comment (evrose54) [reply docs/source/userguide/command_line.rst:93]: Not quite yet. It will after #644.

File: docs/source/userguide/command_line.rst
Comment (evrose54) [reply docs/source/userguide/command_line.rst:185]: No it has. I still think it might be worth just waiting to update the documentation until all CLI PRs are merged in. Up to you though.

File: docs/source/userguide/command_line.rst
Comment (evrose54) [reply docs/source/userguide/command_line.rst:590]: See comment above.

File: docs/source/userguide/command_line.rst
Comment (evrose54) [reply docs/source/userguide/command_line.rst:611]: Yes. See comment above. Happy to change now or in a final PR that updates the documentation.

File: docs/source/releases/latest/627-CLI-tree.yaml
Comment (jsolbrig) [reply docs/source/releases/latest/627-CLI-tree.yaml:32]: Let's update the docs at the end of getting the CLI merged.

Comment (mindyls) [reply docs/source/releases/latest/627-CLI-tree.yaml:32]: Works for me

**Review by mindyls — APPROVED** [review]
Looks good,  docs will be moved later.

File: tests/unit_tests/commandline/test_geoips_tree.py
Comment (evrose54) [reply tests/unit_tests/commandline/test_geoips_tree.py:52]: No, this will fail later on. Can fix once we have all of the other CLI PRs merged. This was wrote at the time that command still existed.

File: geoips/commandline/geoips_tree.py
Comment (evrose54) [reply geoips/commandline/geoips_tree.py:123]: Sounds good!

**Pull request merged**
